panic: no expansion has been registered for ephemeral resource

### Terraform Version

```shell
1.10.0 (371c96c1a126f3cb467e7c59d195abeace13d751)
```


### Terraform Configuration Files

```terraform
terraform {
  required_providers {
    random = {
      source = "hashicorp/random"
    }
  }
}

ephemeral "random_password" "example" {
  length = 12
}

locals {
  test = ephemeral.random_password.example
}
```

### Debug Output

https://gist.github.com/radeksimko/9c4c741a912e05f3bdf5d770051bae03

### Expected Behavior

`terraform console` to not crash

### Actual Behavior

```

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
Please report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version, the stack trace
shown below, and any additional information which may help replicate the issue.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

panic: no expansion has been registered for ephemeral.random_password.example
goroutine 136 [running]:
runtime/debug.Stack()
	/opt/homebrew/Cellar/go/1.23.2/libexec/src/runtime/debug/stack.go:26 +0x64
github.com/hashicorp/terraform/internal/logging.PanicHandler()
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/logging/panic.go:84 +0x184
panic({0x1058f8aa0?, 0x14000797300?})
	/opt/homebrew/Cellar/go/1.23.2/libexec/src/runtime/panic.go:785 +0x124
github.com/hashicorp/terraform/internal/terraform.(*Graph).walk.func1.1()
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/terraform/graph.go:59 +0x36c
panic({0x1058f8aa0?, 0x14000797300?})
	/opt/homebrew/Cellar/go/1.23.2/libexec/src/runtime/panic.go:785 +0x124
github.com/hashicorp/terraform/internal/instances.(*Expander).ResourceInstanceKeys(0x14000ad0858?, {{}, {0x107c05500, 0x0, 0x0}, {{}, 0x45, {0x140008da220, 0xf}, {0x140008da230, ...}}})
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/instances/expander.go:448 +0x1e4
github.com/hashicorp/terraform/internal/terraform.(*evaluationStateData).getEphemeralResource(0x14000a682d0, {{}, 0x45, {0x140008da220, 0xf}, {0x140008da230, 0x7}}, {{0x140008da140, 0x7}, {0xe, ...}, ...})
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/terraform/evaluate.go:810 +0x33c
github.com/hashicorp/terraform/internal/terraform.(*evaluationStateData).GetResource(0x14000a682d0, {{}, 0x45, {0x140008da220, 0xf}, {0x140008da230, 0x7}}, {{0x140008da140, 0x7}, {0xe, ...}, ...})
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/terraform/evaluate.go:580 +0x76c
github.com/hashicorp/terraform/internal/lang.(*Scope).evalContext(0x14000a68360, {0x140002e00a0, 0x1, 0x1}, {0x0, 0x0})
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/lang/eval.go:375 +0x1410
github.com/hashicorp/terraform/internal/lang.(*Scope).EvalContext(...)
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/lang/eval.go:246
github.com/hashicorp/terraform/internal/lang.(*Scope).EvalExpr(0x14000a68360, {0x10603e8d0, 0x140000d3440}, {{0x10603ecb8?, 0x107c05500?}})
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/lang/eval.go:171 +0x8c
github.com/hashicorp/terraform/internal/terraform.(*BuiltinEvalContext).EvaluateExpr(0x105fefd88?, {0x10603e8d0, 0x140000d3440}, {{0x10603ecb8?, 0x107c05500?}}, {0x0?, 0x0?})
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/terraform/eval_context_builtin.go:331 +0x84
github.com/hashicorp/terraform/internal/terraform.evaluateLocalValue(0x0?, {{}, {0x140008da238?, 0x0?}}, {0x14000735c00, 0xa}, {0x106061b38, 0x14000a7e200})
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/terraform/node_local.go:235 +0x43c
github.com/hashicorp/terraform/internal/terraform.(*NodeLocal).Execute(0x140002eff50, {0x106061b38, 0x14000a7e200}, 0xf0?)
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/terraform/node_local.go:145 +0xbc
github.com/hashicorp/terraform/internal/terraform.(*ContextGraphWalker).Execute(0x140003806e0, {0x106061b38, 0x14000a7e200}, {0x12d94adf0, 0x140002eff50})
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/terraform/graph_walk_context.go:161 +0xa4
github.com/hashicorp/terraform/internal/terraform.(*Graph).walk.func1({0x105d6baa0, 0x140002eff50})
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/terraform/graph.go:143 +0x5fc
github.com/hashicorp/terraform/internal/dag.(*Walker).walkVertex(0x14000b08a80, {0x105d6baa0, 0x140002eff50}, 0x1400071f780)
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/dag/walk.go:384 +0x2a0
created by github.com/hashicorp/terraform/internal/dag.(*Walker).Update in goroutine 123
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/dag/walk.go:307 +0xc18
```

### Steps to Reproduce

1. Place custom build of `random` provider w/ ephemeral `random_password` https://github.com/hashicorp/terraform-provider-random/tree/av/ephemeral-password to `~/.terraform.d/plugins/registry.terraform.io/hashicorp/random/1.0.0/{OS}_{ARCH}`
1. `terraform init`
1. `terraform console`


### Additional Context

_No response_

### References

_No response_

--- a/internal/terraform/context_eval_test.go
+++ b/internal/terraform/context_eval_test.go
@@ -332,3 +332,45 @@ func TestContextApplyAndEval(t *testing.T) {
 		}
 	})
 }
+
+func TestContextEval_ephemeralResource(t *testing.T) {
+	// make sure referenced to ephemeral resources are st least valid in the
+	// console, even if they are not known
+	m := testModuleInline(t, map[string]string{
+		"main.tf": `
+ephemeral "ephem_resource" "data" {}
+
+locals {
+  composedString = "prefix-${ephemeral.ephem_resource.data.value}-suffix"
+}
+  `,
+	})
+
+	p := &testing_provider.MockProvider{
+		GetProviderSchemaResponse: &providers.GetProviderSchemaResponse{
+			EphemeralResourceTypes: map[string]providers.Schema{
+				"ephem_resource": {
+					Block: &configschema.Block{
+						Attributes: map[string]*configschema.Attribute{
+							"value": {
+								Type:     cty.String,
+								Computed: true,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	ctx := testContext2(t, &ContextOpts{
+		Providers: map[addrs.Provider]providers.Factory{
+			addrs.NewDefaultProvider("ephem"): testProviderFuncFixed(p),
+		},
+	})
+
+	_, diags := ctx.Eval(m, states.NewState(), addrs.RootModuleInstance, &EvalOpts{
+		SetVariables: testInputValuesUnset(m.Module.Variables),
+	})
+	assertNoErrors(t, diags)
+}

--- a/internal/terraform/evaluate.go
+++ b/internal/terraform/evaluate.go
@@ -629,9 +629,9 @@ func (d *evaluationStateData) GetResource(addr addrs.Resource, rng tfdiags.Sourc
 			}
 
 		default:
-			// We should only end up here during the validate walk,
-			// since later walks should have at least partial states populated
-			// for all resources in the configuration.
+			// We should only end up here during the validate walk (or
+			// console/eval), since later walks should have at least partial
+			// states populated for all resources in the configuration.
 			return cty.DynamicVal, diags
 		}
 	}
@@ -790,8 +790,10 @@ func (d *evaluationStateData) GetResource(addr addrs.Resource, rng tfdiags.Sourc
 func (d *evaluationStateData) getEphemeralResource(addr addrs.Resource, rng tfdiags.SourceRange) (cty.Value, tfdiags.Diagnostics) {
 	var diags tfdiags.Diagnostics
 
-	if d.Operation == walkValidate {
-		// Ephemeral instances are never live during the validate walk
+	if d.Operation == walkValidate || d.Operation == walkEval {
+		// Ephemeral instances are never live during the validate walk. Eval is
+		// similarly offline, and since there is no value stored we can't return
+		// anything other than dynamic.
 		return cty.DynamicVal.Mark(marks.Ephemeral), diags
 	}
 

